apps/examples/testcase/ta_tc/audio/utc/utc_audio_main.c

```diff
@@ -925,6 +922,259 @@ static void utc_audio_pcm_drain_n(void)
+* @apicovered       pcm_wait
+* @precondition     NA
+* @postcondition    NA
+*/
+
+static void utc_audio_pcm_commit_n(void)
+{
+	int ret;
+
+	ret = pcm_mmap_commit(NULL, 0, 0);
+	TC_ASSERT_LT("pcm_mmap_commit", ret, 0);
+
+	g_pcm = pcm_open(0, 0, PCM_IN, NULL);
+	TC_ASSERT_GT("pcm_mmap_begin", pcm_is_ready(g_pcm), 0);
+	ret = pcm_mmap_commit(g_pcm, 0, 0);
+	TC_ASSERT_LT_CLEANUP("pcm_mmap_commit", ret, 0, pcm_close(g_pcm));
+
+	pcm_close(g_pcm);
+	TC_SUCCESS_RESULT();
+}
+
+/**
+* @testcase         audio_pcm_read_p
+* @brief            read predetermined bytes of data from pcm and store in a file
+* @scenario         open pcm in mmap mode and read some data and store it in a file
+* @apicovered       pcm_avail_update, pcm_wait, pcm_mmap_begin, pcm_mmap_commit
+* @precondition     NA
+* @postcondition    NA
+*/
+
+static void utc_audio_pcm_mmap_read_p(void)
+{
+	int fd;
+	char input_str[16];
+	int ret;
+	char *str;
+	unsigned int bytes_per_frame;
+	int remain;
+	int avail;
+	unsigned int frames;
+	unsigned int offset;
+	char *buf;
+
+	fd = open(AUDIO_TEST_FILE, O_RDWR | O_CREAT | O_TRUNC);
+	TC_ASSERT_GEQ_CLEANUP("pcm_mmap_read", fd, 0, pcm_close(g_pcm));
+
+	g_pcm = pcm_open(0, 0, PCM_IN | PCM_MMAP, NULL);
+	TC_ASSERT_GT_CLEANUP("pcm_mmap_read", pcm_is_ready(g_pcm), 0, clean_all_data(fd, NULL));
+
+	bytes_per_frame = pcm_frames_to_bytes(g_pcm, 1);
+	remain = AUDIO_DEFAULT_RATE * AUDIO_RECORD_DURATION;
+
+	printf("Sample will be recorded for 3s, press any key to start(Total frame: %d)\n", remain);
+	fflush(stdout);
+	str = gets(input_str);
+	TC_ASSERT_NEQ_CLEANUP("pcm_mmap_read", str, NULL, clean_all_data(fd, NULL));
+
+	ret = pcm_start(g_pcm);
+	TC_ASSERT_GEQ_CLEANUP("pcm_mmap_read", ret, 0, clean_all_data(fd, NULL));
+
+	while (remain > 0) {
+		avail = pcm_avail_update(g_pcm);
+		TC_ASSERT_GEQ_CLEANUP("pcm_avail_update", avail, 0, clean_all_data(fd, NULL));
+
+		if (avail == 0) {
+			ret = pcm_wait(g_pcm, -1);
+			TC_ASSERT_GEQ_CLEANUP("pcm_wait", ret, 0, clean_all_data(fd, NULL));
+			continue;
+		}
+
+		ret = pcm_mmap_begin(g_pcm, (void **)&buf, &offset, &frames);
+		TC_ASSERT_GEQ_CLEANUP("pcm_mmap_begin", ret, 0, clean_all_data(fd, NULL));
+
+		remain -= frames;
+		ret = write(fd, buf + offset, bytes_per_frame * frames);
+		TC_ASSERT_EQ_CLEANUP("pcm_mmap_read", ret, (bytes_per_frame * frames), clean_all_data(fd, NULL));
+
+		ret = pcm_mmap_commit(g_pcm, offset, frames);
+		TC_ASSERT_GEQ_CLEANUP("pcm_mmap_commit", ret, 0, clean_all_data(fd, NULL));
+	}
+
+	printf("Record done.\n");
+
+	clean_all_data(fd, NULL);
+	TC_SUCCESS_RESULT();
+}
+
+/**
+* @testcase         audio_pcm_write_p
+* @brief            write some predetermined amout of data from file to pcm
+* @scenario         open a pcm with MMAP flag and write data using mmap mode api's
+* @apicovered       pcm_avail_update, pcm_wait, pcm_mmap_begin, pcm_mmap_commit
+* @precondition     NA
+* @postcondition    NA
+*/
+
+static void utc_audio_pcm_mmap_write_p(void)
```

Comment:
You don't use mmap_write anywhere

Reply:
Yes. Actually, I have written this UTC to test the complete mmap write scenario. Just like utc_audio_pcm_writei_p() or utc_audio_pcm_readi_p() which actually performs a proper audio write or read from file. I want to have similar type of API for mmap write / read.

Reply:
Ok I understand your approach but role of utc is that checking open api works fine or not.
so we should call mm_write or mm_read in utc instead of same scenario.

Reply:
Added testcases for all APIs.

Reply:
All these apis should be opened?

Reply:
Yes. We need to open these API's.
avail, wait, begin, commit ---> actual mmap implementation to avoid memcpy
mmap_read, mmap_write ---> to be compatible with original alsa implementation.